chore(upgradeSQLCipher): Upgrading SQLCipher to version 5.4.5

alexjba opened the pull request:
Based on benchmarking data we can get the best performance using the latest `sqlcipher` with the improved crypto implementation and increase the DB page size. (see the chart below)

In this PR the `sqlcipher` version is bumped to latest release (v5.4.5) and the DB connection will be opened using v3 compatibility mode (1024 page size and keeping the improved SHA1 and AES)

## Changes:
### github.com/mutecomm/go-sqlcipher
1. The improved crypto argorighms from go-sqlcipher v3 are merged in v4 
Tags: 
v4.4.2-status.1 - merge `burn_stack` improvement 
v4.4.2-status.2 - merge `SHA1` improvement 
v4.4.2-status.4- merge 'AES' improvement

2. Fixed `go-sqlcipher` to support v3 database in compatibility mode (`sqlcipher` already supports this) (Tag: v4.4.2-status.3)
3. Upgrade `sqlcipher` to v5.4.5 (Tag: v4.5.4-status.1)

### github.com/status-im/migrate/v4
1. Upgrade `go-sqlcipher` version in `github.com/status-im/migrate/v4`

### status-go
1. Upgrade `go-sqlcipher` and `migrate` modules in status-go
2. Configure the DB connections to open the DB in v3 compatibility mode



<img width="1003" alt="Screenshot 2023-05-26 at 15 46 46" src="https://github.com/status-im/status-go/assets/47811206/6f8848ea-2aa6-416b-92fe-58286cf352b0">

alexjba committed:
chore(upgradeSQLCipher): Upgrading SQLCipher to version 5.4.5

Changes:
### github.com/mutecomm/go-sqlcipher
1. The improved crypto argorighms from go-sqlcipher v3 are merged in v4
	Tags:
	v4.4.2-status.1 - merge `burn_stack` improvement
	v4.4.2-status.2 - merge `SHA1` improvement
	v4.4.2-status.4- merge 'AES' improvement
2. Fixed `go-sqlcipher` to support v3 database in compatibility mode (`sqlcipher` already supports this) (Tag: v4.4.2-status.3)
3. Upgrade `sqlcipher` to v5.4.5 (Tag: v4.5.4-status.1)

### github.com/status-im/migrate/v4
1. Upgrade `go-sqlcipher` version in `github.com/status-im/migrate/v4`

### status-go
1. Upgrade `go-sqlcipher` and `migrate` modules in status-go
2. Configure the DB connections to open the DB in v3 compatibility mode

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://status.im/docs/))?
- [ ] Have you tested changes with mobile?
- [ ] Have you tested changes with desktop?

alexjba commented on vendor/golang.org/x/net/context/context.go:
```diff
@@ -1,56 +0,0 @@
-// Copyright 2014 The Go Authors. All rights reserved.
```
@cammellos Not sure why `make vendor` removed `golang.org/x/net/context`. Should I leave it like this?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (11)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | abe3c6eb | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3559/1/) | 2023-05-31 19:18:34 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230531-191507-abe3c6-pr3559.zip) |
| :heavy_check_mark: | abe3c6eb | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3559/1/) | 2023-05-31 19:19:21 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230531-191507-abe3c6-pr3559.zip) |
| :heavy_check_mark: | abe3c6eb | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3559/1/) | 2023-05-31 19:20:18 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230531-191507-abe3c6-pr3559.aar) |
| | | | | | | |
| :heavy_check_mark: | 20a032df | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3559/2/) | 2023-05-31 19:34:28 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230531-193118-20a032-pr3559.zip) |
| :heavy_check_mark: | 20a032df | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3559/2/) | 2023-05-31 19:39:04 | ~7 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230531-193118-20a032-pr3559.zip) |
| :heavy_check_mark: | 20a032df | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3559/2/) | 2023-05-31 19:40:21 | ~9 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230531-193118-20a032-pr3559.aar) |
| :heavy_multiplication_x: | 20a032df | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3559/2/) | 2023-05-31 19:50:19 | ~14 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3559/2//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 722fc619 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3559/3/) | 2023-06-06 07:02:49 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230606-070003-722fc6-pr3559.zip) |
| :heavy_check_mark: | 722fc619 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3559/3/) | 2023-06-06 07:02:53 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230606-070004-722fc6-pr3559.zip) |
| :heavy_check_mark: | 722fc619 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3559/3/) | 2023-06-06 07:05:13 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230606-065958-722fc6-pr3559.aar) |
| :heavy_multiplication_x: | 722fc619 | [#9](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3559/9/) | 2023-06-06 07:21:55 | ~21 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3559/9//consoleText) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_multiplication_x: | f461a1b1 | [#10](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3559/10/) | 2023-06-06 12:23:49 | ~38 sec | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3559/10//consoleText) |
| :heavy_check_mark: | f461a1b1 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3559/4/) | 2023-06-06 12:27:29 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230606-122310-f461a1-pr3559.zip) |
| :heavy_check_mark: | f461a1b1 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3559/4/) | 2023-06-06 12:28:08 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230606-122310-f461a1-pr3559.aar) |
| :heavy_check_mark: | f461a1b1 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3559/4/) | 2023-06-06 12:30:59 | ~7 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230606-122310-f461a1-pr3559.zip) |
| | | | | | | |
| :heavy_check_mark: | d1dc4bc2 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3559/5/) | 2023-06-06 12:37:58 | ~7 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230606-123027-d1dc4b-pr3559.zip) |
| :heavy_check_mark: | d1dc4bc2 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3559/5/) | 2023-06-06 12:38:07 | ~7 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230606-123102-d1dc4b-pr3559.zip) |
| :heavy_multiplication_x: | d1dc4bc2 | [#11](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3559/11/) | 2023-06-06 12:38:11 | ~7 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3559/11//consoleText) |
| :heavy_check_mark: | d1dc4bc2 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3559/5/) | 2023-06-06 12:39:18 | ~8 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230606-123021-d1dc4b-pr3559.aar) |
| :heavy_multiplication_x: | d1dc4bc2 | [#12](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3559/12/) | 2023-06-06 13:06:48 | ~22 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3559/12//consoleText) |
| :heavy_multiplication_x: | d1dc4bc2 | [#13](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3559/13/) | 2023-06-06 13:32:02 | ~21 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3559/13//consoleText) |
| :heavy_multiplication_x: | d1dc4bc2 | [#14](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3559/14/) | 2023-06-06 14:03:55 | ~28 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3559/14//consoleText) |
| :heavy_multiplication_x: | d1dc4bc2 | [#15](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3559/15/) | 2023-06-06 14:36:03 | ~27 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3559/15//consoleText) |
| :heavy_check_mark: | d1dc4bc2 | [#16](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3559/16/) | 2023-06-06 14:40:45 | ~3 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3559/16//consoleText) |

alexjba committed:
chore(upgradeSQLCipher): Use sqlcipher v3 configuration to encrypt a plain text database

alexjba commented on sqlite/sqlite.go:
```diff
@@ -71,6 +71,20 @@ func EncryptDB(unencryptedPath string, encryptedPath string, key string, kdfIter
 		return err
 	}
 
+	if _, err := db.Exec("PRAGMA encrypted.cipher_page_size = 1024"); err != nil {
```
@cammellos how can I test this flow? I see it's not used in StatusDesktop

osmaczko reviewed (approved):
Cool!

alexjba committed:
chore(upgradeSQLCipher): Scanning NULL BLOB value should return nil

Fixing failing tests: TestSyncDeviceSuite/TestPairingSyncDeviceClientAsReceiver;  TestSyncDeviceSuite/TestPairingSyncDeviceClientAsSender

Considering the following configuration:
1. Table with BLOB column has 1 NULL value
2. Query the value
3. Rows.Scan(&dest sql.NullString)

Expected: dest.Valid == false; dest.String == nil
Actual: dest.Valid == true; dest.String == ""

alexjba committed:
chore: Bump go-sqlcipher version to include NULL BLOB fix

alexjba commented:
Updated the PR to fix failing tests.
The failure was caused by a go-sqlcipher bug where querying a null blob value in a `sql.NullString` would result in a NullString that is valid and not null.

This is the fix:
https://github.com/status-im/status-go/pull/3559/commits/f461a1b1a72ceca0655beafae5bf15c18711a566

alexjba commented:
> Updated the PR to fix failing tests. The failure was caused by a go-sqlcipher bug where querying a null blob value in a `sql.NullString` would result in a NullString that is valid and not null.
> 
> This is the fix: [f461a1b](https://github.com/status-im/status-go/commit/f461a1b1a72ceca0655beafae5bf15c18711a566)

Now there are other failures (sometimes different failures on different runs) that can't be reproduced locally:
FAIL: TestMessengerBackupSuite (15.04s)
FAIL    github.com/status-im/status-go/wakuv2    19.679s

I hope the reliability issue is not introduced by this PR
